app/src/main/java/org/simple/clinic/util/ErrorResolver.kt

```diff
@@ -0,0 +1,43 @@
+package org.simple.clinic.util
+
+import io.reactivex.exceptions.CompositeException
+import okhttp3.internal.http2.ConnectionShutdownException
+import okhttp3.internal.http2.StreamResetException
+import java.net.SocketException
+import java.net.SocketTimeoutException
+import java.net.UnknownHostException
+
+object ErrorResolver {
+
+  fun resolve(error: Throwable): ResolvedError {
+    val actualCause = findActualCause(error)
+
+    return when (actualCause) {
+      is SocketException,
+      is SocketTimeoutException,
+      is UnknownHostException,
+      is StreamResetException,
+      is ConnectionShutdownException -> ResolvedError.NetworkRelated(actualCause)
+      else -> ResolvedError.Other(actualCause)
+    }
+  }
+
+  private fun findActualCause(error: Throwable): Throwable {
+    var actualError = error
+    if (actualError is CompositeException) {
+      // CompositeException wraps the cause inside another
+      // exception called CompositeExceptionCausalChain.
+      actualError = findActualCause(actualError.cause.cause!!)
+    }
+    // This place may identify more wrapped errors
+    // like UndeliverableException, etc. in the future.
+    return actualError
+  }
+}
+
+sealed class ResolvedError(val actualCause: Throwable, val shouldBeReported: Boolean) {
+
+  class NetworkRelated(actualCause: Throwable) : ResolvedError(actualCause, shouldBeReported = false)
```

Comment:
I don't think `shouldBeReported` should be a property of the `ResolvedError` type. The reason is that the responsibility of this class is to classify errors into network related errors or not. Whether they should be reported or not must be the responsibility of the reporter.

Reply:
Makes sense, I agree with this. Will change this.